Orange/widgets/unsupervised/owlouvainclustering.py

```diff
@@ -0,0 +1,377 @@
+from collections import deque
+from concurrent.futures import Future  # pylint: disable=unused-import
+from enum import Enum
+
+import networkx as nx  # pylint: disable=unused-import
+import numpy as np
+from AnyQt.QtCore import Qt, pyqtSignal as Signal, QObject
+from AnyQt.QtWidgets import QSlider, QCheckBox, QWidget  # pylint: disable=unused-import
+from types import SimpleNamespace as namespace
+from typing import Optional  # pylint: disable=unused-import
+
+from Orange.clustering.louvain import table_to_knn_graph, Louvain
+from Orange.data import Table, DiscreteVariable
+from Orange.projection import PCA
+from Orange.widgets import widget, gui, report
+from Orange.widgets.settings import DomainContextHandler, ContextSetting, \
+    Setting
+from Orange.widgets.utils.annotated_data import get_next_name, add_columns, \
+    ANNOTATED_DATA_SIGNAL_NAME
+from Orange.widgets.utils.concurrent import ThreadExecutor
+from Orange.widgets.utils.signals import Input, Output
+from Orange.widgets.widget import Msg
+
+try:
+    from orangecontrib.network.network import Graph
+except:
+    Graph = None
+
+
+_MAX_PCA_COMPONENTS = 50
+_DEFAULT_PCA_COMPONENTS = 25
+_MAX_K_NEIGBOURS = 200
+_DEFAULT_K_NEIGHBORS = 30
+
+
+METRICS = [('Euclidean', 'l2'), ('Manhattan', 'l1')]
+
+
+class TaskQueue(QObject):
+    """Not really a task queue `per-se`. Running start will run the tasks in
+    the current list and cannot handle adding other tasks while running."""
+    on_exception = Signal(Exception)
+    on_complete = Signal()
+    on_progress = Signal(float)
+    on_cancel = Signal()
+
+    def __init__(self, parent=None):
+        super().__init__(parent=parent)
+        self.__tasks = deque()
+        self.__progress = 0
+        self.__cancelled = False
+
+    def cancel(self):
+        self.__cancelled = True
+
+    def push(self, task):
+        self.__tasks.append(task)
+
+    def __set_progress(self, progress):
+        # Only emit progress signal when the progress has changed sufficiently
+        if int(progress * 100) > int(self.__progress * 100):
+            self.on_progress.emit(progress)
+        self.__progress = progress
+
+    def start(self):
+        num_tasks = len(self.__tasks)
+
+        for idx, task_spec in enumerate(self.__tasks):
+
+            if self.__cancelled:
+                self.on_cancel.emit()
+                return
+
+            def __task_progress(percentage):
+                current_progress = idx / num_tasks
+                # How much progress can each task contribute to the total
+                # work to be done
+                task_percentage = len(self.__tasks) ** -1
```

Comment:
What is wrong with 1 / len(self.__tasks)? :)

Reply:
I think I wrote this code at a point where I was doing a considerable amount of calculus and I guess I found this clearer at that point.